Optimize the listener close process

akutz opened the pull request:
This patch refactors how listeners are closed with relationship to pending I/O. While the provider must track connections in order to prevent a duplicate address from listening on a specified network, the remote endpoints that spawn from these connections need not have unique addresses. To this end a listener no longer cares about tracking active connections and the connection doesn't need to remove itself from the tracker. Instead an event system with a done channel is used so that when a listener is closed all of its pending I/O is unblocked and closed immediately.

cc @dougm

akutz commented:
Hi @dougm,

I CCd you in the description because I cannot seem to [request a review from you](https://help.github.com/articles/requesting-a-pull-request-review/) less I make you a collaborator, and since I'm not an org it appears I can only make collaborators that have push access?

akutz commented:
Hi @dougm,

Yep: 

> Note: In a private repository, repository owners can only grant write access to collaborators. Collaborators can't have read-only access to repositories owned by a user account.

dougm commented on memconn_listener.go:
```diff
@@ -87,24 +51,28 @@ func (l *listener) dial(
 	return localWrapper, nil
 }
 
+type errListenerClosed struct{}
```
Could you just use `errors.New` instead of these new error types?
```go
var errListenerClosed = errors.New("closed")
```
Could also be useful to callers if exported?
```go
if err == memconn.ErrListenerClosed {
   // ignoring closed
}

akutz replied:
Hi @dougm,

I personally think comparing errors like that is fraught with potential for failure due to the way errors get wrapped, rereferenced , dereferenced, etc. I prefer:

```
if err != nil && err.Error() == memconn.ErrListenerClosed.Error() {
```

Regardless, I made it a struct because it follows the pattern in other Go code I've seen, takes up zero memory unless used (aside from the constant), and can easily have special functions added to it that makes it adhere to error-related interfaces (see the `net` pkg for more info).

The other reason *not* to do `type errListenerClosed string` and then assign it to a constant string is because you're forced to type assert when you return the value from the `Error() string` function.

dougm replied:
Sure, I think if wrapped, etc., consider that a different error.  The 2 error types in this PR in their current form are similar to `io.EOF` as an example, and the concise use of `if err == io.EOF` can keep things simpler.  Though this kind of use case may not apply to these error types.

akutz replied:
Hi @dougm,

My point is that even with `io.EOF` I don't believe it is safe to do a direct equality test since you don't necessarily know if the error will be wrapped or if, for example, some returned `errors.New` with the same value that is contained by `io.EOF`.

akutz committed:
Optimize the listener close process

This patch refactors how listeners are closed with relationship to
pending I/O. While the provider must track connections in order to
prevent a duplicate address from listening on a specified network, the
remote endpoints that spawn from these connections need not have unique
addresses. To this end a listener no longer cares about tracking active
connections and the connection doesn't need to remove itself from the
tracker. Instead an event system with a done channel is used so that
when a listener is closed all of its pending I/O is unblocked and closed
immediately.

dougm commented on memconn_pipe_go110.go:
```diff
@@ -14,3 +14,12 @@ import (
 	return net.Pipe()
 }
+
+func isClosedChan(c <-chan struct{}) bool {
```
Could use `context.Context` instead of the bare chan, then `ctx.Done()` instead of `isClosedChan`.  Might also be useful for callers to make use of Context to set deadlines, cancel, etc. as it is for a tcp conn.

akutz replied:
Hi @dougm,

Yep. I actually started down that path the first time. I walked it back because:

1. Keep it simple, keep it fast
2. I'm trying to make `memconn` at first satisfy `net.Conn`, not necessarily all the features of something like `TCPConn`, etc.

In some future update I'll likely add a custom `Dialer` like `net.Dialer` that recognizes Memconn networks and implement `context` then since the `net.Dialer` uses the context.

dougm reviewed (approved):
lgtm

I haven't looked at the rest of memconn yet, so feel free to disregard my comments as I don't have much context.
